pyro/infer/mcmc/adaptation.py

```diff
@@ -194,25 +178,394 @@ def step(self, t, z, accept_prob):
+            r_flat = torch.cat([r[site_name].reshape(-1) for site_name in site_names])
+            u[site_names] = _matmul(mass_matrix_sqrt_inverse, r_flat)
+        return u
+
+
+def _arrowhead_sqrt(x):
+    assert isinstance(x, ArrowheadMatrix)
+    head_size = x.top.size(0)
+    if head_size == 0:
+        return ArrowheadUpperSqrtMatrix(x.top, x.bottom_diag.sqrt())
+
+    A, B = x.top[:, :head_size], x.top[:, head_size:]
+    # NB: the complexity is O(N * head_size^2)
+    # ref: https://en.wikipedia.org/wiki/Schur_complement#Background
+    Dsqrt = x.bottom_diag.sqrt()
+    B_Dsqrt = B / Dsqrt.unsqueeze(-2)  # shape: head_size x N
+    schur_complement = A - B_Dsqrt.matmul(B_Dsqrt.t())  # complexity: head_size^2 x N
+    # we will decompose schur_complement to U @ U.T (so that the sqrt matrix
+    # is upper triangular) using some `flip` operators:
+    #   flip(cholesky(flip(schur_complement)))
+    top_left = torch.flip(torch.flip(schur_complement, (-2, -1)).cholesky(), (-2, -1))
+    top_right = B_Dsqrt
+    top = torch.cat([top_left, top_right], -1)
+    bottom_diag = Dsqrt
+    return ArrowheadUpperSqrtMatrix(top, bottom_diag)
+
+
+def _arrowhead_sqrt_to_sqrt_inverse(x):
+    assert isinstance(x, ArrowheadUpperSqrtMatrix)
+    head_size = x.top.size(0)
+    if head_size == 0:
+        return ArrowheadUpperSqrtMatrix(x.top, x.bottom_diag.reciprocal())
+
+    A, B = x.top[:, :head_size], x.top[:, head_size:]
+    B_Dinv = B / x.bottom_diag.unsqueeze(-2)
+
+    identity = torch.eye(head_size, dtype=A.dtype, device=A.device)
+    top_left = torch.triangular_solve(identity, A, upper=True)[0]
+    top_right = -top_left.matmul(B_Dinv)  # complexity: head_size^2 x N
+    top = torch.cat([top_left, top_right], -1)
+    bottom_diag = x.bottom_diag.reciprocal()
+    return ArrowheadUpperSqrtMatrix(top, bottom_diag)
+
+
+def _arrowhead_sqrt_matmul(x, y, transpose=False):
+    assert isinstance(x, ArrowheadUpperSqrtMatrix)
+    head_size = x.top.size(0)
+    if transpose:
+        z = x.top.transpose(-2, -1).matmul(y[:head_size])
+        # here we exploit the diagonal structure of the bottom right part
+        # of arrowhead_sqrt matrix; so the complexity is still O(N)
+        top = z[:head_size]
+        bottom = z[head_size:] + x.bottom_diag * y[head_size:]
+    else:
+        top = x.top.matmul(y)
+        bottom = x.bottom_diag * y[head_size:]
+    return torch.cat([top, bottom], 0)
+
+
+def _arrowhead_sqrt_inverse_to_inverse(x):
+    assert isinstance(x, ArrowheadUpperSqrtMatrix)
+    head_size = x.top.size(0)
+    if head_size == 0:
+        return x.bottom_diag.pow(2)
+
+    A, B = x.top[:, :head_size], x.top[:, head_size:]
+    top = A.t().matmul(x.top)
+    bottom_left = top[:, head_size:].t()
+    # the following matmul operator is O(N^2 x head_size)
+    bottom_right = B.t().matmul(B) + x.bottom_diag.pow(2).diag()
+    return torch.cat([top, torch.cat([bottom_left, bottom_right], -1)], 0)
+
+
+class ArrowheadMassMatrix:
```

Comment:
@fritzo This mass matrix adapter has complexity O(matrix_size x head_size^2) in time and O(matrix_size x head_size) in memory. So the performance should be similar to the diagonal mass matrix when head_size << matrix_size. It is nice to try on the SIR models. There are a few features we can add easily (which I can add in the follow-up PRs if needed) such as:
+ a pack/unpack mechanism to move important parts of a latent variable to the head part of the arrowhead matrix. I think this is the feature that you requested in #2425.
+ specify the initial mass matrix (which is learned using SVI) in the constructor